### DISCUSSION TITLE
[4.4] How to properly inject a synthetic service from within a bundle?

### DISCUSSION THREAD

**Discussion by addiks** [symfony/symfony#51113] [Q&A]
Hello,

I am currently trying to inject the composer class-loader (returned by the `composer/autoload.php`) into the symfony service container because I need it for some class-to-filepath resolution in one of my bundles.

As far as I understand, I am supposed to define the service as a synthetic service and then inject it into the container (A). But **where** exactly am I supposed to inject the service into the container then? As a bundle-distributor, I cannot modify the kernel of the projects where my bundle will be used, so the most likely place to make the injection would be in my bundle's `boot` method.

The problem with injecting it inside the `boot` method of my bundle is that the symfony kernel will build the container before booting the bundles (B), so symfony will complain about the missing synthetic service before I have any chance injecting it (at least during the `cache:clear` command).

How do I solve this? 
Are synthetic services not meant to be used in bundles?
Thank you for any help you can provide.

A: https://symfony.com/doc/current/service_container/synthetic_services.html
B: https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/HttpKernel/Kernel.php#L136

```
$ php bin/console cache:clear -v

 // Clearing the cache for the prod environment with debug false                                                        

 // Clearing outdated warmup directory...                                                                               

 // Warming up cache...                                                                                                 


In Container.php line 272:
                                                                                                            
  [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]                                
  The "composer.class_loader" service is synthetic, it needs to be set at boot time before it can be used.  
                                                                                                            

Exception trace:
  at /usr/workspace/MyCompany/MySoftware/vendor/symfony/dependency-injection/Container.php:272
 Symfony\Component\DependencyInjection\Container->make() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/dependency-injection/Container.php:231
 Symfony\Component\DependencyInjection\Container->get() at /usr/workspace/MyCompany/MySoftware/var/cache/pro_/ContainerAizHDOO/getAddiksRdm_BlackMagickCacheWarmerService.php:9
 require() at /usr/workspace/MyCompany/MySoftware/var/cache/pro_/ContainerAizHDOO/srcMyCompany_MySoftware_KernelProdContainer.php:942
 ContainerAizHDOO\srcMyCompany_MySoftware_KernelProdContainer->load() at /usr/workspace/MyCompany/MySoftware/var/cache/pro_/ContainerAizHDOO/getCacheWarmerService.php:20
 ContainerAizHDOO\srcMyCompany_MySoftware_KernelProdContainer->{closure}() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/http-kernel/CacheWarmer/CacheWarmerAggregate.php:89
 Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerAggregate->warmUp() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/http-kernel/Kernel.php:631
 Symfony\Component\HttpKernel\Kernel->initializeContainer() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/http-kernel/Kernel.php:136
 Symfony\Component\HttpKernel\Kernel->boot() at /usr/workspace/MyCompany/MySoftware/src/Kernel.php:55
 MyCompany\MySoftware\Kernel->boot() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/http-kernel/Kernel.php:153
 Symfony\Component\HttpKernel\Kernel->reboot() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:189
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->warmup() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/framework-bundle/Command/CacheClearCommand.php:129
 Symfony\Bundle\FrameworkBundle\Command\CacheClearCommand->execute() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/console/Application.php:1039
 Symfony\Component\Console\Application->doRunCommand() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/framework-bundle/Console/Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/console/Application.php:275
 Symfony\Component\Console\Application->doRun() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/framework-bundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /usr/workspace/MyCompany/MySoftware/vendor/symfony/console/Application.php:149
 Symfony\Component\Console\Application->run() at /usr/workspace/MyCompany/MySoftware/bin/console:52

cache:clear [--no-warmup] [--no-optional-warmers] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

**Comment by MatTheCat** [comment]
Hello,

> I need [...] some class-to-filepath resolution in one of my bundles

I would have thought this requires a custom `autoload` section in the bundle's `composer.json`?

FTR you're supposed to configure the service as synthetic before injecting it, so you would have to follow https://symfony.com/doc/4.4/bundles/extension.html. But I feel like your problem is a X/Y one, not something to be solved like this.

**Reply by addiks** [reply]
> I would have thought this requires a custom autoload section in the bundle's composer.json?

How does an `autoload` section in the `composer.json` give me a component that I can ask during runtime to provide me with the file-path to a not-yet-loaded class? I feel like you misunderstand what I am looking for. I don't want to add autoloading for my own classes, I want something that can resolve classnames to filepaths, like the `findFile` method of the `ClassLoader` object that the `vendor/autoload.php` returns.

> FTR you're supposed to configure the service as synthetic before injecting it, so you would have to follow https://symfony.com/doc/4.4/bundles/extension.html.

I **have** configured it as a synthetic service, otherwise symfony would not complain about it being a synthetic service that was never injected. The problem is not the service-configuration, the problem is that I have no unproblematic place inside the bundle where the injection can happen.

> But I feel like your problem is a X/Y one, not something to be solved like this.

What is an X/Y Problem?

**Reply by MatTheCat** [reply]
> What is an X/Y Problem?

An X/Y problem is when you encounter an issue (Y) when trying to fix another (X), and you ask for Y instead of X.

> I want something that can resolve classnames to filepaths, like the `findFile` method of the `ClassLoader` object that the `vendor/autoload.php` returns.

If Composer knows about these classes, [`ReflectionClass::getFileName`](https://www.php.net/manual/en/reflectionclass.getfilename.php) should do the job?

> The problem is not the service-configuration, the problem is that I have no unproblematic place inside the bundle where the injection can happen.

If the service has properly been configured, you should be able to set the instance in your bundle’s `boot` method :thinking:

**Reply by addiks** [reply]
> If Composer knows about these classes, [ReflectionClass::getFileName](https://www.php.net/manual/en/reflectionclass.getfilename.php) should do the job?

This requires that the class was already loaded into the PHP runtime, which would be too late. I need to load a modified version of the class into runtime.

> If the service has properly been configured, you should be able to set the instance in your bundle’s boot method

No, that does not work because symfony builds the container in the `initializeContainer` method (and thus expects the synthetic service to be already set) **before** calling the `boot` method of the bundle. It fails because of the missing service before I have any chance of injecting it.

**Reply by MatTheCat** [reply]
> I need to load a modified version of the class into runtime.

Hmmm, monkey-patching, uh?

> No, that does not work because symfony builds the container in the `initializeContainer` method (and thus expects the synthetic service to be already set) before calling the `boot` method of the bundle.

AFAIU it does not expect the service to be **set** but **configured**. I managed to use a synthetic service from a bundle in a personal project so it is definitely possible.

**Reply by MatTheCat** [reply]
Okay looking at the stack trace I see your service is tagged as `kernel.cache_warmer` which is why it is created before your bundle’s `boot` method.

I still have no idea why you would do such a thing, but wouldn’t it be better to register a custom autoloader while setting [Composer’s `prepend-autoloader`](https://getcomposer.org/doc/06-config.md#prepend-autoloader) to `false`?
